kettle-plugins/hadoop-cluster/ui/src/main/java/org/pentaho/big/data/kettle/plugins/hadoopcluster/ui/dialog/wizard/util/NamedClusterHelper.java

```diff
@@ -0,0 +1,197 @@
+        }
+      }
+      List<CachedFileItemStream> fileItemStreams =
+        copyAndUnzip( fileInputStream, FileType.CONFIGURATION, siteFile.getName(), fileName, manager );
+      for ( CachedFileItemStream cachedFileItemStream : fileItemStreams ) {
+        siteFiles.put( cachedFileItemStream.getFieldName(), cachedFileItemStream );
+      }
+    }
+    return siteFiles;
+  }
+
+  public static List<CachedFileItemStream> copyAndUnzip( InputStream fileInputStream, FileType fileType,
+                                                         String fileName, String realFileName,
+                                                         HadoopClusterManager manager )
+    throws IOException {
+    List<CachedFileItemStream> unzippedFileItemStreams = new ArrayList<>();
+    if ( realFileName.endsWith( ".zip" ) ) {
+      try ( ZipInputStream zis = new ZipInputStream( fileInputStream ) ) {
+        for ( ZipEntry zipEntry = zis.getNextEntry(); zipEntry != null; zipEntry = zis.getNextEntry() ) {
+          if ( !zipEntry.isDirectory() ) {
+            //remove all directory structure from the zip file names and only unzip the files
+            String[] split = zipEntry.getName().split( "/" ); //zip files always use forward slash
+            String unzippedFileName = split[ split.length - 1 ];
+            if ( isValidUpload( unzippedFileName, fileType, manager ) ) {
+              CachedFileItemStream unzippedFileItemStream =
+                new CachedFileItemStream( zis, unzippedFileName, unzippedFileName );
+              unzippedFileItemStream.setLastModified( zipEntry.getLastModifiedTime().toMillis() );
+              unzippedFileItemStreams.add( unzippedFileItemStream );
+            }
+          }
+        }
+      }
+    } else {
+      //file is not zipped
+      if ( isValidUpload( realFileName, fileType, manager ) ) {
+        unzippedFileItemStreams.add( new CachedFileItemStream( fileInputStream, fileName,
+          realFileName ) );
+      }
+    }
+    return unzippedFileItemStreams;
+  }
+
+  private static boolean isValidUpload( String fileName, FileType fileType, HadoopClusterManager manager ) {
+    boolean valid = ( fileType.equals( FileType.CONFIGURATION ) && manager.isValidConfigurationFile( fileName ) )
+      ||
+      ( fileType.equals( FileType.DRIVER ) && fileName.endsWith( FileType.DRIVER.getValue() ) );
+    return valid;
+  }
+
+  public static boolean processDriverFile( String driverFile, HadoopClusterManager manager ) throws Exception {
+    boolean result = false;
+    File file = new File( driverFile );
+    FileInputStream driverStream = new FileInputStream( file );
+    if ( isValidUpload( file.getName(), FileType.DRIVER, manager ) ) {
+      String destination = Const.getShimDriverDeploymentLocation();
+      String fileSeparator = System.getProperty( "file.separator" );
+      FileUtils.copyInputStreamToFile( driverStream,
+        new File( destination + fileSeparator + file.getName() ) );
```

Comment:
[comment]: <> (FrogbotReviewComment)


## 🎯 Static Application Security Testing (SAST) Vulnerability
<div align='center'>

| Severity                | Finding                  |
| :---------------------: | :-----------------------------------: |
| ![](https://raw.githubusercontent.com/jfrog/frogbot/master/resources/v2/applicableMediumSeverity.png)<br>  Medium | Stored Path Traversal |

</div>

<details>
<summary> <b>Full description</b> </summary>
<br>


### Overview
Path traversal, also known as directory traversal, is a type of
vulnerability that allows an attacker to access files or directories on a
computer or device that are outside of the intended directory.
Allowing arbitrary read access can allow the attacker to read sensitive
files, such as configuration files or sensitive data, potentially leading
data loss or even system compromise. Allowing arbitrary write access is
more severe and in most cases leads to arbitrary code execution, via
editing important system files or sensitive data.

### Vulnerable example
```java
public class path_traversaLvuln {
    protected void doGet(HttpServletRequest request, HttpServletResponse response)
            throws ServletException, IOException {
        String DOCS_FOLDER = "/srv/www/docs";
        String docName = statement.executeQuery(query);  // Reading from DB
        Path docPath = Paths.get(DOCS_FOLDER, docName);
        File docFile = docPath.toFile();
        FileUtils.copyFile(docFile, response.getOutputStream());
    }
}
```
In this example, an attacker can, via a stored parameter, inject a back-path,
that will get anywhere in the system, using "../../".

### Remediation
```diff
public class path_traversal_safe {
    protected void doGet(HttpServletRequest request, HttpServletResponse response)
            throws ServletException, IOException {
        String DOCS_FOLDER = "/srv/www/docs";
        String docName = statement.executeQuery(query);  // Reading from DB
        Path docPath = Paths.get(DOCS_FOLDER, docName);
        + Path normDocPath = docPath.normalize();
        + // Make sure the canonical path resides in the desired dir
        + if (normDocPath.startsWith(DOCS_FOLDER)) {
            File docFile = docPath.toFile();
            FileUtils.copyFile(docFile, response.getOutputStream());
        + }
    }
}
```
By checking that the folder name still starts with the predefined prefix, we
make sure that the attacker is not able to back-path outside of the allowed
folder.


</details>

<details>
<summary> <b>Code Flows</b> </summary>
<br>


<details>
<summary> <b>Vulnerable data flow analysis result</b> </summary>
<br>


↘️ `System.getProperty( "file.separator" )` (at kettle-plugins/hadoop-cluster/ui/src/main/java/org/pentaho/big/data/kettle/plugins/hadoopcluster/ui/dialog/wizard/util/NamedClusterHelper.java line 190)

↘️ `fileSeparator` (at kettle-plugins/hadoop-cluster/ui/src/main/java/org/pentaho/big/data/kettle/plugins/hadoopcluster/ui/dialog/wizard/util/NamedClusterHelper.java line 190)

↘️ `destination + fileSeparator` (at kettle-plugins/hadoop-cluster/ui/src/main/java/org/pentaho/big/data/kettle/plugins/hadoopcluster/ui/dialog/wizard/util/NamedClusterHelper.java line 192)

↘️ `destination + fileSeparator + file.getName()` (at kettle-plugins/hadoop-cluster/ui/src/main/java/org/pentaho/big/data/kettle/plugins/hadoopcluster/ui/dialog/wizard/util/NamedClusterHelper.java line 192)

↘️ `new File( destination + fileSeparator + file.getName() )` (at kettle-plugins/hadoop-cluster/ui/src/main/java/org/pentaho/big/data/kettle/plugins/hadoopcluster/ui/dialog/wizard/util/NamedClusterHelper.java line 192)


</details>


</details>


---
<div align='center'>

[🐸 JFrog Frogbot](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>